### PR TITLE
JournalDB -> Box<JournalDB>, and it's a trait.

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: util/src/journaldb.rs [hunk 1]
```diff
@@ -31,22 +47,12 @@ use std::env;
 /// write operations out to disk. Unlike OverlayDB, `remove()` operations do not take effect
 /// immediately. Rather some age (based on a linear but arbitrary metric) must pass before
 /// the removals actually take effect.
-pub struct JournalDB {
+pub struct OptionOneDB {
```
Comment (tomusdrw) [inline util/src/journaldb.rs]: Is it possible to give it some more descriptive name?

File: ethcore/src/client/client.rs [hunk 2]
```diff
@@ -139,8 +139,8 @@ impl<V> Client<V> where V: Verifier {
 		state_path.push("state");
 
 		let engine = Arc::new(try!(spec.to_engine()));
-		let mut state_db = JournalDB::from_prefs(state_path.to_str().unwrap(), config.prefer_journal);
-		if state_db.is_empty() && engine.spec().ensure_db_good(&mut state_db) {
+		let mut state_db = Box::new(OptionOneDB::from_prefs(state_path.to_str().unwrap(), config.prefer_journal));
```
Comment (tomusdrw) [inline ethcore/src/client/client.rs]: Small factory for those DB will be needed like `JournalDB::new(JorunalDBType::OptionOne)`

File: util/src/overlaydb.rs [hunk 3]
```diff
@@ -36,6 +36,7 @@ use kvdb::{Database};
 ///
 /// `lookup()` and `contains()` maintain normal behaviour - all `insert()` and `remove()` 
 /// queries have an immediate effect in terms of these functions.
+#[derive(Clone)]
```
Comment (tomusdrw) [inline util/src/overlaydb.rs]: Is it needed?

File: ethcore/src/client/client.rs (hunk 2 above)
Comment (gavofyork) [inline ethcore/src/client/client.rs]: later PR for that.

File: util/src/journaldb.rs (hunk 1 above)
Comment (gavofyork) [inline util/src/journaldb.rs]: The algorithm's name is "option one".

**Pull request merged**